tensorboard/uploader/uploader_subcommand_test.py

```diff
@@ -115,6 +115,44 @@ def testUploadIntentUnderDryRunOneShot(self):
             mock_stdout_write.call_args_list[-1][0][0], "\nDone.\n"
         )
 
+    def testUploadIntentWithExperimentUrlCallback(self):
+        """Test the upload intent with a callback."""
+        mock_server_info = mock.MagicMock()
+        mock_channel = mock.MagicMock()
+        upload_limits = server_info_pb2.UploadLimits(
+            max_scalar_request_size=128000,
+            max_tensor_request_size=128000,
+            max_tensor_point_size=11111,
+            max_blob_request_size=128000,
+            max_blob_size=128000,
+        )
+        tb_dev_url = "https://tensorboard.dev/experiment/test_experiment_id"
+        mock_experiment_url_callback = mock.MagicMock()
+        mock_stdout_write = mock.MagicMock()
+        with mock.patch.object(
```

Comment:
So much stuff is being mocked out here that I have no idea what code
paths under test are actually being tested.

Can we just use real values instead? `mock_server_info` shouldn’t be a
mock; it should be a `ServerInfoResponse` proto value. Ditto for
`upload_limits`, but you shouldn’t need that at all. Patching the
library functions in `server_info_lib` seems crazy to me—that’s real
code that we want to test, and it doesn’t even have side effects. Also,
please use `Mock` instead of `MagicMock` unless you specifically need
the magic methods, which you don’t here. (See internal “Python Tip of
the Week #049: Mocking pitfalls”.)

How about something like (sketch; adjust to taste):

```python
    def testUploadIntentWithExperimentUrlCallback(self):
        """Test the upload intent with a callback."""
        server_info = server_info_pb2.ServerInfoResponse()
        server_info.url_format.template = "https://tensorboard.dev/x/{}"
        server_info.url_format.id_placeholder = "{}"

        stub = dry_run_stubs.DryRunTensorBoardWriterStub()
        stub.CreateExperiment = lambda req, **__: write_service_pb2.CreateExperimentResponse(
            experiment_id="test_experiment_id", url="this URL is ignored"
        )

        expected_url = "https://tensorboard.dev/x/test_experiment_id"

        with mock.patch.object(
            dry_run_stubs, "DryRunTensorBoardWriterStub", wraps=lambda: stub,
        ), mock.patch.object(sys.stdout, "write"):
            mock_channel = mock.Mock()
            mock_experiment_url_callback = mock.Mock()
            intent = uploader_subcommand.UploadIntent(
                self.get_temp_dir(),
                dry_run=True,
                one_shot=True,
                experiment_url_callback=mock_experiment_url_callback,
            )
            intent.execute(server_info, mock_channel)
        mock_experiment_url_callback.assert_called_once_with(expected_url)
```

Reply:
Yeah happy to do all that, but I don't have bandwidth to similarly clean up all the other tests here from which I pasted the structure.  Local consistency is also a virtue, right?

Reply:
Sure, understood that it’s copy-pasted, and not asking you to clean up
the others. Local consistency is a virtue *ceteris paribus*, but in this
case (imho) the ceteris seem rather far from paribus, and so the
aphorism that springs more readily to my mind is to not let the perfect
be the enemy of the good. But I do welcome your thoughts—let me know if
you disagree with the direction that I’m suggesting.

Reply:
> the ceteris seem rather far from paribus

*Chef's kiss*

Since I like to paste things, I pasted your sketch.  Worked first time and looks correct to me.  So, thanks for ghostwriting this whole PR!